pkg/operator/controller/externaldns/rbac_test.go

```diff
@@ -0,0 +1,625 @@
+						Name: controller.ExternalDNSBaseName,
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							APIGroups: []string{"extensions", "networking.k8s.io"},
+							Resources: []string{"ingresses"},
+							Verbs:     []string{"get", "list", "watch"},
+						},
+						{
+							APIGroups: []string{""},
+							Resources: []string{"endpoints", "services", "pods"},
+							Verbs:     []string{"get", "list", "watch"},
+						},
+					},
+				},
+			},
+			expectedExist: true,
+			expectedRole: rbacv1.ClusterRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: controller.ExternalDNSBaseName,
+				},
+				Rules: []rbacv1.PolicyRule{
+					{
+						APIGroups: []string{"extensions", "networking.k8s.io"},
+						Resources: []string{"ingresses"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+					{
+						APIGroups: []string{""},
+						Resources: []string{"endpoints", "services", "pods", "nodes"},
+						Verbs:     []string{"get", "list", "watch"},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			cl := fake.NewFakeClient(tc.existingObjects...)
+			r := &reconciler{
+				client: cl,
+				log:    zap.New(zap.UseDevMode(true)),
+			}
+			gotExist, gotRole, _ := r.ensureExternalDNSClusterRole(context.TODO())
+			if gotExist != tc.expectedExist {
+				t.Errorf("expected cluster roles's exist to be %t, got %t", tc.expectedExist, gotExist)
+			}
+			if reflect.DeepEqual(*gotRole, tc.expectedRole) {
+				t.Errorf("expected cluster role %v, got %v", tc.expectedRole, gotRole)
+			}
+		})
+	}
+}
+
+func TestExternalDNSRoleRulesChanged(t *testing.T) {
+	testCases := []struct {
+		name          string
+		inputCurrent  []rbacv1.PolicyRule
+		inputExpected []rbacv1.PolicyRule
+		expectChanged bool
+	}{
+		{
+			name: "Same",
+			inputCurrent: []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{"extensions", "networking.k8s.io"},
+					Resources: []string{"ingresses"},
+					Verbs:     []string{"get", "list", "watch"},
+				},
+				{
+					APIGroups: []string{""},
+					Resources: []string{"endpoints", "services", "pods", "nodes"},
+					Verbs:     []string{"get", "list", "watch"},
+				},
+			},
+			inputExpected: []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{"extensions", "networking.k8s.io"},
+					Resources: []string{"ingresses"},
+					Verbs:     []string{"get", "list", "watch"},
+				},
+				{
+					APIGroups: []string{""},
+					Resources: []string{"endpoints", "services", "pods", "nodes"},
+					Verbs:     []string{"get", "list", "watch"},
+				},
+			},
+			expectChanged: false,
+		},
+		{
+			name: "All reordered",
+			inputCurrent: []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{"extensions", "networking.k8s.io"},
+					Resources: []string{"ingresses"},
+					Verbs:     []string{"get", "list", "watch"},
+				},
+				{
+					APIGroups: []string{"extensions"},
+					Resources: []string{"ingresses"},
+					Verbs:     []string{"get", "watch", "list"},
+				},
```

Comment:
This rule isn't even in the `inputExpected` value; is that intentional?  This is more than just "reordered".

Reply:
The idea was that even if all is reordered the content is the same. I changed the test name to make it clearer. Also I removed the `extentions/ingress` resources as per your previous comments.